File: src/main/java/uk/co/ben_gibson/git/link/UI/Action/Menu/MenuAction.java
package uk.co.ben_gibson.git.link.UI.Action.Menu;

import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.editor.*;
import com.intellij.openapi.fileEditor.FileEditorManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import uk.co.ben_gibson.git.link.UI.Action.Action;
import uk.co.ben_gibson.git.link.UI.LineSelection;

import java.util.List;

/**
 * An action triggered from the view or right click menu.
 */
abstract class MenuAction extends Action
{
    protected abstract void perform(@NotNull final Project project, @NotNull final VirtualFile file, @Nullable final LineSelection lineSelection);

    @Override
    public void actionPerformed(final Project project, @NotNull final AnActionEvent event)
    {
        VirtualFile file = event.getData(CommonDataKeys.VIRTUAL_FILE);

        if (file == null || project == null) {
            return;
        }

        Editor editor = FileEditorManager.getInstance(project).getSelectedTextEditor();

        LineSelection lineSelection = null;

        if (editor != null) {
            lineSelection = this.getLineSelection(editor);
        }

        this.perform(project, file, lineSelection);
    }

    @NotNull
    private LineSelection getLineSelection(@NotNull final Editor editor)
    {
        CaretModel caretModel        = editor.getCaretModel();
        List<CaretState> caretStates = caretModel.getCaretsAndSelections();

        if (caretStates.size() < 1) {
            return new LineSelection(caretModel.getLogicalPosition().line + 1);
        }

        CaretState caretState = caretStates.get(0);
        LogicalPosition start = caretState.getSelectionStart();
        LogicalPosition end   = caretState.getSelectionEnd();

        if (start == null || end == null) {
            return new LineSelection(caretModel.getLogicalPosition().line + 1);
        }

        return new LineSelection(start.line + 1, end.line + 1);
    }

    @Override
    protected boolean shouldActionBeEnabled(@NotNull final AnActionEvent event)
    {
        return (event.getData(CommonDataKeys.VIRTUAL_FILE) != null);
    }
}

File: src/main/kotlin/uk/co/ben_gibson/git/link/ui/actions/menu/MenuAction.kt
package uk.co.ben_gibson.git.link.ui.actions.menu

import com.intellij.openapi.actionSystem.ActionPlaces
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.project.Project
import uk.co.ben_gibson.git.link.*
import uk.co.ben_gibson.git.link.git.LineSelection
import uk.co.ben_gibson.git.link.ui.actions.Action
import uk.co.ben_gibson.git.link.editor.lineSelection

abstract class MenuAction(type: Type) : Action(type) {
    override fun buildContext(project: Project, event: AnActionEvent): Context? {
        val file = event.getData(CommonDataKeys.VIRTUAL_FILE) ?: return null

        return Context.File(file, resolveLineSelection(event))
    }

    private fun resolveLineSelection(event: AnActionEvent): LineSelection? {
        val editor = event.getData(CommonDataKeys.EDITOR) ?: return null

        // Honour an explicit selection anywhere, but only fall back to the caret line when the action came
        // from the editor itself, not from the main menu, a project view entry or an editor tab
        if (!editor.selectionModel.hasSelection() && event.place in WHOLE_FILE_PLACES) {
            return null
        }

        return editor.lineSelection
    }

    companion object {
        private val WHOLE_FILE_PLACES = setOf(
            ActionPlaces.MAIN_MENU,
            ActionPlaces.PROJECT_VIEW_POPUP,
            ActionPlaces.EDITOR_TAB_POPUP,
        )
    }
}
